File: src/main/scala/fish/philwants/modules/TwitterModule.scala
package fish.philwants.modules

import fish.philwants.Credentials
import org.jsoup.Connection.Response

object TwitterModule extends AbstractModule {
  val uri = "https://twitter.com/"
  val moduleName = "Twitter"

  /**
   * The Twitter login process uses a hidden form parameter `authenticity_token` and requires
   * a session cookie. Visit the login page, persist the cookies, submit the form with the hidden
   * parameter
   */
  def tryLogin(creds: Credentials): Boolean = {
    val resp = get(uri).execute()

    // Get the form and update it with credentials
    val form = resp.selectForm("form.LoginForm.js-front-signin")
      .update("session[username_or_email]", creds.username)
      .update("session[password]", creds.password)

    // Try to login
    val loginResp = form
      .submit()
      .cookies(resp.cookies())
      .header("Content-Type", "application/x-www-form-urlencoded")
      .followRedirects(false)
      .execute()

    // Check login result
    isLoginSuccessful(loginResp)
  }

  def isLoginSuccessful(resp: Response): Boolean = {
    val locationHeader = resp.header("Location")
    !locationHeader.contains("login/error?username_or_email")
  }
}

File: secondary_implementations/python/modules/TwitterModule.py
# vim: ts=2 sw=2 noexpandtab
import requests
from lxml import html
from .AbstractModule import AbstractModule


class TwitterModule(AbstractModule):
	def try_login(self, uname, password):
		agent = AbstractModule().user_agent() 
		headers = {'user-agent': agent}
		resp = requests.get('https://twitter.com/', headers=headers)

		tree = html.fromstring(resp.content)
		payload = tree.forms[4].fields
		payload = {k : v for k, v in payload.items()}
		payload['session[username_or_email]'] = uname
		payload['session[password]'] = password
		headers = {
			'user-agent': agent,
			'Content-Type': 'application/x-www-form-urlencoded'}
		loginResp = requests.post(
			'https://twitter.com/sessions', data=payload, headers=headers,
			cookies=resp.cookies, allow_redirects=False)

		return 'login/error?username_or_email' not in loginResp.headers['Location']


twitter = TwitterModule()
